app/models/reservation.rb

```diff
@@ -0,0 +1,22 @@
+class Reservation < ActiveRecord::Base
+  belongs_to :request
+  has_one :donor
+
+  state_machine :initial => :unreserved do
+    event :reserve do
+      transition :unreserved => :reserved
+    end
+
+    event :ship do
+      transition :reserved => :shipped
+    end
+
+    event :receive do
+      transition :shipped => :received
+    end
+
+    event :cancel do
+      transition :reserved => :unreserved, :shipped => :reserved
```

Comment:
generally i think the `:unreserved` state is unnecessary - when a reservation in the `:reserved` state is canceled, we can just delete it. that also avoids the problem of an event handling two transitions, which is a footgun waiting to happen